source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,138 @@
+
+public struct EnrollmentData{
+    /// K: UTXO key, A public key matching a frozen UTXO
+    public PublicKey K;
+
+    /// X: random seed, The nth image of random value
+    public Hash X;
+
+    /// n: cycle length, A number within bounds [0; 1008]
+    /// original spec: [0; (Balance(K) % 40,000) * freezing period * 2]
+    public ulong n;
+
+    /// R: signature noise, The initial nonce used for signing
+    public Point R;
+
+    /// S: A signature for the messge H(K, X, n, R) and the key K, using R
+    public Signature S;
+
+    /***************************************************************************
+        Implements hashing support
+        Params:
+            dg = hashing function
+    ***************************************************************************/
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        hashPart(this.K, dg);
+        hashPart(this.X, dg);
+        hashPart(this.n, dg);
+        hashPart(this.R, dg);
+    }
+}
+
+/*******************************************************************************
+    Handles data and process for enrollment
+*******************************************************************************/
+
+public struct EnrollmentManager
+{
+    /// Node's key pair
+    private KeyPair node_keypair;
+
+    /// Random key for enrollment
+    private Pair random_key;
+
+    /// Node balance
+    private Amount balance;
+
+    // Random object
+    private Random genint;
+
+    /// EnrollmentData
+    private EnrollmentData* data;
+
+    /***************************************************************************
+        Constructor
+        Params:
+            keypair = the keypair of the owner node
+            balance = account balance of the owner node
+    ***************************************************************************/
+    public this(KeyPair keypair, Amount balance)
+    {
+        this.node_keypair = keypair;
+        this.balance = balance;
+        this.genint = Random(123);
+        this.data = null;
+    }
+
+    /***************************************************************************
+        Makes a data object for enrollment process
+        Returns:
+            a EnrollmentData object
+    ***************************************************************************/
+    public EnrollmentData getEnrollmentData()
+    {
+        this.data = new EnrollmentData();
+        
+        // public key
+        this.data.K = this.node_keypair.address;
+
+        // cycle length
+        this.data.n = uniform(0, 1008, this.genint);
+
+        // random seed
+        this.data.X = hashFull(this.node_keypair.secret.data);
+        round(this.data.X, this.data.n - 1);
+
+        // signature noise
+        this.random_key = Pair.random();
+        this.data.R = this.random_key.V;
+
+        // signature
+        Hash hash = hashFull(*this.data);
+        this.data.S = this.node_keypair.secret.sign(hash[]);
```

Comment:
The signature part actually has to be a schnorr signature using `r` (the `random_key.v` scalar) above. If the enrollment is not signed with `r`, then you can't prove that you actually know `R`. :)

https://github.com/bpfkorea/agora/blob/5ef14fe0ce879da52cc47716d7688bf5ab0d943f/source/agora/common/crypto/Schnorr.d#L196

^ This function can be used. It's currently private, but I'll make it public in a PR in a moment.